### PR TITLE
Switch response metric

### PR DESCRIPTION
**Pull request by lawhead**
# Overview

Added an Inquiry Preview session metric for switch response time.

## Ticket

https://www.pivotaltracker.com/story/show/180605653

## Contributions

- Added trigger types for triggers used in the Inquiry Preview / Copy Phrase task.
- Added new metric that computes the average switch response time in seconds.
- Some refactoring of the trigger helpers to correctly load existing triggers.txt files. 
- Unit tests for new functionality

## Test

- Ran all unit tests.
- Ran copy phrase task to ensure that all session data was correctly written.
- Tested the switch_response_time calculations against previously generated triggers.txt files using different `preview_inquiry_progress_method` values.

### CODE REVIEW

File: bcipy/helpers/triggers.py [hunk 1]
```diff
@@ -651,7 +651,10 @@ class TriggerType(Enum):
     TARGET = "target"
     FIXATION = "fixation"
     PROMPT = "prompt"
-    SYSTEM = "offset"
+    SYSTEM = "offset_correction"
```
Comment (tab-cmd) [inline bcipy/helpers/triggers.py]: I have a PR to change how we write and read system offsets. It will be as implemented (with a single offset written). I think it'd be best to wait for that PR. I should be done this week.

File: bcipy/helpers/triggers.py [hunk 2]
```diff
@@ -651,7 +651,10 @@ class TriggerType(Enum):
...
+    CALIBRATION = "calib"
+    KEY_PRESS = "key_press"
```
Comment (tab-cmd) [inline bcipy/helpers/triggers.py]: I'd be in favor of a more generic `EVENT` or similar and having the label describe what it is... key press, mouse click, etc. What do you think? It would constitute a change in the task as well. I'm happy to address that in another PR if you're okay with this approach.

File: bcipy/helpers/triggers.py [hunk 3]
```diff
@@ -784,33 +793,13 @@ def load(path: str,
...
-                            float(trigger[2])
-                            )
-                )
-            except Exception as e:
-                raise BciPyCoreException(f'{trigger} read from {path} is not able to be cast')
-
-        return new_trigger_list
+        triggers, system_offset = TriggerHandler.read_text_file(path)
+        excluded_types = exclusion or []
+        total_offset = offset + system_offset
```
Comment (tab-cmd) [inline bcipy/helpers/triggers.py:314]: I like how succinct this is, however the custom exceptions were helpful and intentional. I'd be okay changing the logic flow but we need some error handling with better messaging.

File: bcipy/task/paradigm/rsvp/copy_phrase.py [hunk 4]
```diff
@@ -715,6 +722,55 @@ def btn_press_count(self) -> int:
...
+        """Computes the average switch response in seconds."""
+
+        # Remove consecutive items with the same type; we are only interested
+        # in PREVIEW followed by a KEY_PRESS.
+        triggers = destutter(self.switch_triggers(), key=lambda trg: trg.type)
+        pairs = list(zip(triggers[::2], triggers[1::2]))
+
+        # Confirm that the data is structured as expected.
+        for preview, keypress in pairs:
+            valid = preview.type == TriggerType.PREVIEW and keypress.type == TriggerType.KEY_PRESS
```
Comment (tab-cmd) [inline bcipy/task/paradigm/rsvp/copy_phrase.py]: This is hard to follow visually.. maybe

```
valid = (preview.type == TriggerType.PREVIEW) and (keypress.type == TriggerType.KEY_PRESS)
```

File: bcipy/task/paradigm/rsvp/copy_phrase.py [hunk 5]
```diff
@@ -715,6 +722,55 @@ def btn_press_count(self) -> int:
...
+        if not self.trigger_path:
+            return []
+        triggers, _offset = TriggerHandler.read_text_file(self.trigger_path)
+        return [
+            trg for trg in triggers
+            if trg.type in [TriggerType.PREVIEW, TriggerType.KEY_PRESS]
+        ]
+
+
+def destutter(items: List, key: Callable = lambda x: x) -> List:
```
Comment (tab-cmd) [inline bcipy/task/paradigm/rsvp/copy_phrase.py]: This should live somewhere else... helpers? I can suggest something if one doesn't come to mind.

**Review by tab-cmd — CHANGES_REQUESTED** [review]

File: bcipy/helpers/triggers.py (hunk 1 above)
Comment (lawhead) [reply bcipy/helpers/triggers.py]: Okay, sounds good. If possible, I would like to support any previously used trigger types.

File: bcipy/helpers/triggers.py (hunk 3 above)
Comment (lawhead) [reply bcipy/helpers/triggers.py:314]: The custom exceptions that we previously had here were swallowing some of the errors I was encountering which made it hard to debug. I went ahead and added them back (see the `read_text_file` method) with some additional details to hopefully make debugging easier.

File: bcipy/helpers/triggers.py (hunk 2 above)
Comment (lawhead) [reply bcipy/helpers/triggers.py]: I added the `EVENT` prefix to the keypress for now, but I think this is not quite what you had in mind? Let's discuss in our next meeting and create a followup task for this.

File: bcipy/task/paradigm/rsvp/copy_phrase.py (hunk 5 above)
Comment (lawhead) [reply bcipy/task/paradigm/rsvp/copy_phrase.py]: This function seemed a bit too generic for any of our existing helpers so I went ahead and created a new one. If you would rather see it elsewhere let me know.

File: bcipy/helpers/triggers.py (hunk 1 above)
Comment (tab-cmd) [reply bcipy/helpers/triggers.py]: We discussed this a bit in the Trigger PR but the implications are pretty far reaching. As of 2.0, we discussed dropping support for these older offset correction methods. If requested, we could provide a helper to convert older data to this newer format. This would require reading in the triggers.txt and writing a single offset value to the triggers file.

File: bcipy/helpers/list_processing.py [hunk 6]
```diff
@@ -0,0 +1,20 @@
+"""Utility functions for list processing."""
```
Comment (tab-cmd) [inline bcipy/helpers/list_processing.py]: Would `helpers/list.py` work? It might make the module generic enough to pull other operations on lists here.

**Review by tab-cmd — APPROVED** [review]

File: bcipy/task/paradigm/rsvp/copy_phrase.py [hunk 7]
```diff
@@ -744,6 +751,37 @@ def btn_press_count(self) -> int:
...
+        return sum(response_times) / count if count > 0 else None
+
+    def switch_triggers(self) -> List[Trigger]:
+        """Returns a list of switch-related triggers"""
+        if not self.trigger_path:
+            return []
+        triggers, _offset = TriggerHandler.read_text_file(self.trigger_path)
+        return [
+            trg for trg in triggers
+            if trg.type in [TriggerType.PREVIEW, TriggerType.EVENT]
```
Comment (lawhead) [inline bcipy/task/paradigm/rsvp/copy_phrase.py:782]: This is currently working and gives us the appropriate triggers to use for the switch response annotation, but if we add any other triggers that use `TriggerType.EVENT` we need to remember to make the changes here to look for the symbol of interest.
